src/validation.cpp

```diff
@@ -29,10 +39,5412 @@
+                if (plTxnReplaced)
+                    plTxnReplaced->push_back(it->GetSharedTx());
+            }
+            pool.RemoveStaged(allConflicting, false);
+
+            // This transaction should only count for fee estimation if it isn't a
+            // BIP 125 replacement transaction (may not be widely supported), the
+            // node is not behind, and the transaction is not dependent on any other
+            // transactions in the mempool.
+            bool validForFeeEstimation = !fReplacementTransaction && IsCurrentForFeeEstimation() && pool.HasNoInputsOf(tx);
+
+            // Store transaction in memory
+            pool.addUnchecked(hash, entry, setAncestors, validForFeeEstimation);
+
+            // Add memory address index
+            if (fAddressIndex) {
+                pool.addAddressIndex(entry, view);
+            }
+
+            // Add memory spent index
+            if (fSpentIndex) {
+                pool.addSpentIndex(entry, view);
+            }
+
+            if (tx.IsZerocoinSpend()) {
+                pool.countZCSpend++;
+            }
+
+            // trim mempool and check if tx was trimmed
+            if (!fOverrideMempoolLimit) {
+                LimitMempoolSize(pool, GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000, GetArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY) * 60 * 60);
+                if (!pool.exists(hash))
+                    return state.DoS(0, false, REJECT_INSUFFICIENTFEE, "mempool full");
+            }
+        } 
+        else {
+            LockPoints lp;
+            double fSpendsCoinbase = false;
+            CAmount inChainInputValue = 0;
+            CAmount nFees = 0;
+            int64_t nSigOpsCost = GetLegacySigOpCount(tx);
+            CTxMemPool::setEntries setAncestors;
+            CTxMemPoolEntry entry(ptx, nFees, GetTime(), chainActive.Height(),
+                                inChainInputValue, fSpendsCoinbase, nSigOpsCost, lp);
+            pool.addUnchecked(hash, entry, setAncestors, !IsInitialBlockDownload());
+            if (tx.IsZerocoinSpend()) {
+                pool.countZCSpend++;
+            }
+        }
+    }
+
+    if ((tx.IsZerocoinSpend() || tx.IsZerocoinRemint()) && markZcoinSpendTransactionSerial)
+        zcState->AddSpendToMempool(zcSpendSerials, hash);
+    if (tx.IsSigmaSpend()){
+        if(markZcoinSpendTransactionSerial)
+            sigmaState->AddSpendToMempool(zcSpendSerialsV3, hash);
+        LogPrintf("Updating mint tracker state from Mempool..");
+#ifdef ENABLE_WALLET
+        if (zwalletMain) {
+            LogPrintf("Updating spend state from Mempool..");
+            zwalletMain->GetTracker().UpdateSpendStateFromMempool(zcSpendSerialsV3);
+        }
+#endif
+    }
+    if(markZcoinSpendTransactionSerial)
+        sigmaState->AddMintsToMempool(zcMintPubcoinsV3);
+#ifdef ENABLE_WALLET
+    if(tx.IsSigmaMint()){
+        BOOST_FOREACH(const CTxOut &txout, tx.vout)
```

Comment:
This is not critical thing, but looks we already added pubcoins upper, no need to do this again.